docs/guides/validation-user-guide.md

```diff
@@ -4,4 +4,497 @@ headline: Validation User Guide
+something more than just a data structure. One of the commodities required for describing domain
+specifics is data validation. Spine builds on top of Protobuf with a Validation library.
+
+The purpose of the Validation library is:
+ - giving framework users an easy way to maintain simple domain invariants;
+ - giving framework users an easy way to construct simple Anticorruption layers.
+
+This guide will walk you though the API of Spine Validation library. All of the validation features
+described here are currently supported in the Java environment. Many are supported in Dart as well.
+For more info, see the description of individual constraints.
+
+## Validation options
+### Required fields
+
+In Protobuf 2 the concept of required fields used to be built into the language. This proved to be
+a [design mistake](https://stackoverflow.com/a/31814967/3183076) and in Protobuf 3 all the fields
+are optional. However, we've revived the concept in Spine Validation library. Unlike the Protobuf 2
+`required` fields, our `(required)` fields do not effect message serialization and thus can be
+ignored. Rather than guarding a technical invariant, `(required)` guards a domain model invariant.
+The choice whether or not to use validation lies solely on the developer.
+
+#### How required fields work
+
+Fields in Protobuf may have either a primitive type or a user-defined type. A user-defined type is
+a `message` or an `enum` and primitive types are numbers, `string`, and `bytes`. Due to limitations
+of the binary format, there is no way to tell if a numeric field is set to `0` or not just not set.
+This means that a numeric field cannot be required, as there is no way to check if it is set. All 
+the other fields can be required. For `message` fields this means that the message must not be
+empty. For `enum` fields, this means that the enum value must have a number other than `0` (since 
+the enum value with number `0` is the default value of the field). For `string` and `bytes` fields
+this means that the sequence must not be empty.
+
+For collection fields (i.e. `repeated` and `map`), a field is considered set if:
+  1. The collection is not empty.
+  2. At least one of the entries (values for `map`s) matches the rules described above.
+
+Note that collections of numeric fields can be required. In those cases, only the rule 1. applies
+and the rule 2. is ignored.
+
+#### Declaring required fields
+
+In the basic scenario, a single required field is marked with the `(required)` option:
+
+```proto
+import "spine/options.proto";
+
+// A phone number represented by a string of digits.
+message PhoneNumber {
+    string digits = 1 [(required) = true];
+}
+```  
+
+Here, the field `PhoneNumber.digits` is required. If the API user tries to validate an instance of
+`PhoneNumber` without this field, a `ConstraintViolation` is produced:
+
+```java
+PhoneNumber.newBuilder()
+           .setDigits("")
+           .vBuild(); // ← Throws ValidationException.
```

Comment:
I think we should explicitly tell about `vBuild()` and its difference from `build()`.

We describe the API, and this method is surely its part.